<!DOCTYPE html><html lang="en" data-default-color-scheme="auto"><head><script>let sw=navigator.serviceWorker;sw.ready.then(()=>{sw.controller||location.reload()}),sw.register("/sw.js",{scope:"/"})</script><script src="/theme-inject/cdn.js"></script><script defer src="https://api.limour.top/vue/0d2f95c1-755d-436b-adf8-eee12a80ed32/script.js"></script><meta charset="UTF-8"><link rel="apple-touch-icon" sizes="76x76" href="https://img.limour.top/2023/08/29/64ee07361815a.webp"><link rel="icon" href="https://img.limour.top/2023/08/29/64ee07361815a.webp"><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=5,shrink-to-fit=no"><meta http-equiv="x-ua-compatible" content="ie=edge"><meta name="theme-color" content="#2f4154"><meta name="author" content="Limour"><meta name="keywords" content=""><meta name="description" content="FASTA(.fa) 储存参考数据集 从零开始完整学习全基因组测序（WGS）数据分析：第2节 FASTA和FASTQ  基本单元 序列所表示的基因名：&amp;gt;ENSMUSG00000020122ENSMUST00000138518，后可接空格表示注释前缀 具体序列信息：CCCTCCTATCATGC……GGGCCCACCTGTTCTCTGGT 基因名独占一行，序列信息为基因名后一行至下一个 &amp;gt;"><title>【迁移】二代测序数据处理之数据格式说明 - Limour&#39;s Blog</title><link rel="stylesheet" href="/cdn/bootstrap/css/bootstrap.min.css"><link rel="stylesheet" href="/cdn/github-markdown-css/github-markdown.min.css"><link rel="stylesheet" href="/cdn/hint.css/hint.min.css"><link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css"><link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_lbnruvf0jn.css"><link rel="stylesheet" href="/css/main.css"><link id="highlight-css" rel="stylesheet" href="/css/highlight.css"><link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css"><link rel="stylesheet" href="/theme-inject/custom.css"><link rel="stylesheet" href="/theme-inject/iconfont.css"><script id="fluid-configs">var dntVal,Fluid=window.Fluid||{},CONFIG=(Fluid.ctx=Object.assign({},Fluid.ctx),{hostname:"hexo.limour.top",root:"/",version:"1.9.8",typing:{enable:!1,typeSpeed:70,cursorChar:"_",loop:!1,scope:[]},anchorjs:{enable:!0,element:"h1,h2,h3,h4,h5,h6",placement:"left",visible:"hover",icon:"§"},progressbar:{enable:!0,height_px:3,color:"#29d",options:{showSpinner:!1,trickleSpeed:100}},code_language:{enable:!0,default:"TEXT"},copy_btn:!0,image_caption:{enable:!0},image_zoom:{enable:!1,img_url_replace:["",""]},toc:{enable:!0,placement:"right",headingSelector:"h1,h2,h3,h4,h5,h6",collapseDepth:0},lazyload:{enable:!0,loading_img:"https://img.limour.top/inload.svg",onlypost:!1,offset_factor:.5},web_analytics:{enable:!1,follow_dnt:!0,baidu:null,google:{measurement_id:null},tencent:{sid:null,cid:null},leancloud:{app_id:null,app_key:null,server_url:null,path:"window.location.pathname",ignore_local:!1},umami:{src:null,website_id:null,domains:null,start_time:"2024-01-01T00:00:00.000Z",token:null,api_server:null},woyaola:null,cnzz:null},search_path:"/local-search.xml",include_content_in_search:!0});CONFIG.web_analytics.follow_dnt&&(dntVal=navigator.doNotTrack||window.doNotTrack||navigator.msDoNotTrack,Fluid.ctx.dnt=dntVal&&(dntVal.startsWith("1")||dntVal.startsWith("yes")||dntVal.startsWith("on")))</script><script src="/js/utils.js"></script><script src="/js/color-schema.js"></script><link rel="canonical" href="https://hexo.limour.top/er-dai-ce-xu-shu-ju-chu-li-zhi-shu-ju-ge-shi-shuo-ming"><meta name="generator" content="Hexo 7.1.1"><link rel="alternate" href="/atom.xml" title="Limour's Blog" type="application/atom+xml">
<link rel="alternate" href="/rss2.xml" title="Limour's Blog" type="application/rss+xml">
</head><body><header><div class="header-inner" style="height:70vh"><nav id="navbar" class="navbar fixed-top navbar-expand-lg navbar-dark scrolling-navbar"><div class="container"><a class="navbar-brand" href="/"><strong>Limour&#39;s Blog</strong> </a><button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation"><div class="animated-icon"><span></span><span></span><span></span></div></button><div class="collapse navbar-collapse" id="navbarSupportedContent"><ul class="navbar-nav ml-auto text-center"><li class="nav-item"><a class="nav-link" href="https://hexo.limour.top/" target="_self"><i class="iconfont icon-home-fill"></i> <span>Home</span></a></li><li class="nav-item"><a class="nav-link" href="/archives/" target="_self"><i class="iconfont icon-archive-fill"></i> <span>Archive1</span></a></li><li class="nav-item"><a class="nav-link" href="https://occdn.limour.top/archives/" target="_self"><i class="iconfont icon-archive-fill"></i> <span>Archive2</span></a></li><li class="nav-item"><a class="nav-link" href="https://b.limour.top/archives/" target="_self"><i class="iconfont icon-archive-fill"></i> <span>Archive3</span></a></li><li class="nav-item"><a class="nav-link" href="https://od.limour.top/" target="_self"><i class="iconfont icon-onedrive"></i> <span>Alist</span></a></li><li class="nav-item"><a class="nav-link" href="https://orcid.org/0000-0001-8897-1685" target="_self"><i class="iconfont icon-orcid"></i> <span>Orcid</span></a></li><li class="nav-item"><a class="nav-link" href="/links/" target="_self"><i class="iconfont icon-link-fill"></i> <span>Links</span></a></li><li class="nav-item"><a class="nav-link" href="/atom.xml" target="_self"><i class="iconfont icon-rss"></i> <span>RSS</span></a></li><li class="nav-item" id="search-btn"><a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search"><i class="iconfont icon-search"></i></a></li><li class="nav-item" id="color-toggle-btn"><a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle"><i class="iconfont icon-dark" id="color-toggle-icon"></i></a></li></ul></div></div></nav><div id="banner" class="banner" parallax="true" style="background:url(https://img.limour.top/2023/08/29/64ee08e108638.webp) no-repeat center center;background-size:cover"><div class="full-bg-img"><div class="mask flex-center" style="background-color:rgba(0,0,0,.3)"><div class="banner-text text-center fade-in-up"><div class="h2"><span id="subtitle">【迁移】二代测序数据处理之数据格式说明</span></div><div class="mt-3"><span class="post-meta mr-2"><i class="iconfont icon-author" aria-hidden="true"></i> Limour </span><span class="post-meta"><i class="iconfont icon-date-fill" aria-hidden="true"></i> <time datetime="2022-01-17 11:33" pubdate>January 17, 2022 am</time></span></div><div class="mt-1"><span class="post-meta mr-2"><i class="iconfont icon-chart"></i> 2k words </span><span class="post-meta mr-2"><i class="iconfont icon-clock-fill"></i> 18 mins</span></div></div></div></div></div></div></header><main><div class="container-fluid nopadding-x"><div class="row nomargin-x"><div class="side-col d-none d-lg-block col-lg-2"></div><div class="col-lg-8 nopadding-x-md"><div class="container nopadding-x-md" id="board-ctn"><div id="board"><article class="post-content mx-auto"><h1 id="seo-header">【迁移】二代测序数据处理之数据格式说明</h1><p id="updated-time" class="note note-info">Last updated on March 19, 2024 pm</p><div class="markdown-body"><h2 id="FASTA-fa-储存参考数据集">FASTA(.fa) 储存参考数据集</h2><p><a href="https://hexo.limour.top/go/#aHR0cHM6Ly96aHVhbmxhbi56aGlodS5jb20vcC8yODQ3MDg4Mw==" rel="noopener external nofollow noreferrer">从零开始完整学习全基因组测序（WGS）数据分析：第2节 FASTA和FASTQ</a></p><ul><li>基本单元</li><li>序列所表示的基因名：<code>&gt;ENSMUSG00000020122ENSMUST00000138518</code>，后可接空格表示注释前缀</li><li>具体序列信息：<code>CCCTCCTATCATGC……GGGCCCACCTGTTCTCTGGT</code></li><li>基因名独占一行，序列信息为基因名后一行至下一个 <code>&gt;</code> 基因名标记前</li></ul><figure class="highlight txt"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs txt">&gt;ENSMUSG00000020122ENSMUST00000138518<br>CCCTCCTATCATGCTGTCAGTGTATCTCTAAATAGCACTCTCAACCCCCGTGAACTTGGT<br>TATTAAAAACATGCCCAAAGTCTGGGAGCCAGGGCTGCAGGGAAATACCACAGCCTCAGT<br>TCATCAAAACAGTTCATTGCCCAAAATGTTCTCAGCTGCAGCTTTCATGAGGTAACTCCA<br>GGGCCCACCTGTTCTCTGGT<br></code></pre></td></tr></table></figure><ul><li>FASTA文件为基本单元的简单罗列</li></ul><figure class="highlight txt"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><code class="hljs txt">&gt;ENSMUSG00000020122ENSMUST00000138518<br>CCCTCCTATCATGCTGTCAGTGTATCTCTAAATAGCACTCTCAACCCCCGTGAACTTGGT<br>TATTAAAAACATGCCCAAAGTCTGGGAGCCAGGGCTGCAGGGAAATACCACAGCCTCAGT<br>TCATCAAAACAGTTCATTGCCCAAAATGTTCTCAGCTGCAGCTTTCATGAGGTAACTCCA<br>GGGCCCACCTGTTCTCTGGT<br>&gt;……<br>……<br>&gt;ENSMUSG00000020122ENSMUST00000125984<br>GAGTCAGGTTGAAGCTGCCCTGAACACTACAGAGAAGAGAGGCCTTGGTGTCCTGTTGTC<br>TCCAGAACCCCAATATGTCTTGTGAAGGGCACACAACCCCTCAAAGGGGTGTCACTTCTT<br>CTGATCACTTTTGTTACTGTTTACTAACTGATCCTATGAATCACTGTGTCTTCTCAGAGG<br>CCGTGAACCACGTCTGCAAT<br>&gt;……<br>……<br></code></pre></td></tr></table></figure><h2 id="FASTQ-fq-储存原始测序数据">FASTQ(.fq) 储存原始测序数据</h2><ul><li><strong>每四行成为一个独立的单元</strong>，<strong>称之为read</strong>；FASTQ文件为read的简单罗列</li><li>第一行：以‘@’开头，是这一条read的唯一标识符</li><li>第二行：测序read的序列，由A，C，G，T和N这五种字母构成，N代表的是测序时那些无法被识别出来的碱基；</li><li>第三行：以‘+’开头，用以兼容旧版格式</li><li>第四行：测序read的质量值，Q = -10log(测序错误率)，字符=<code>chr(ord('!')+Q)</code>，上限为 <code>~</code></li></ul><figure class="highlight txt"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs txt">@DJB775P1:248:D0MDGACXX:7:1202:12362:49613<br>TGCTTACTCTGCGTTGATACCACTGCTTAGATCGGAAGAGCACACGTCTGAA<br>+<br>JJJJJIIJJJJJJHIHHHGHFFFFFFCEEEEEDBD?DDDDDDBDDDABDDCA<br></code></pre></td></tr></table></figure><h2 id="GTF-gtf-描述基因和转录本的信息">GTF(.gtf) 描述基因和转录本的信息</h2><p><a href="https://hexo.limour.top/go/#aHR0cHM6Ly9jbG91ZC50ZW5jZW50LmNvbS9kZXZlbG9wZXIvYXJ0aWNsZS8xNjI1MjA0" rel="noopener external nofollow noreferrer">GTF文件格式简介</a></p><ul><li>头部有 <code>#</code> 开头的注释行</li><li>主体为 <code>\t</code> 分隔的具有九列的表格，空值用 <code>.</code> 填充</li><li>第一列 <code>seqid</code> 代表染色体的ID</li><li>第二列是 <code>source</code> 代表基因结构的来源</li><li>第三列是feature, 代表区间对应的特征类型，如外显子等</li><li>第四、五列为区间的起止坐标</li><li>第六列是 <code>score</code></li><li>第七列是 <code>strand</code>, 代表正负链的信息, +表示正链，-表示负链，?表示不清楚</li><li>第八列是 <code>phase</code>，当描述的是CDS区间信息时，需要指定翻译时开始的位置，取值范围有0,1,2两种</li><li>第九列是attributes, 表示属性，键值对间以分号分隔，键值对内以空格分隔</li></ul><figure class="highlight txt"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs txt">#!genome-build GRCh38.p12<br>#!genome-version GRCh38<br>#!genome-date 2013-12<br>#!genome-build-accession NCBI:GCA_000001405.27<br>#!genebuild-last-updated 2018-01<br>1       ensembl_havana  gene    65419   71585   .       +       .       gene_id &quot;ENSG00000186092&quot;; gene_version &quot;6&quot;; gene_name &quot;OR4F5&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;;<br></code></pre></td></tr></table></figure><h2 id="IDX-idx-基因组比对工具HISAT2的索引文件">IDX(.idx) 基因组比对工具HISAT2的索引文件</h2><p><a href="https://hexo.limour.top/go/#aHR0cDovL2Jsb2cuYmlvY2hlbi5jb20vYXJjaGl2ZXMvMzM3" rel="noopener external nofollow noreferrer">RNA-Seq基因组比对工具HISAT2</a></p><ul><li>使用 hisat2-build 工具从.fa文件建立</li></ul><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><code class="hljs shell">export FADIR=/opt/human_grch38/dna<br>export FANAME=Homo_sapiens.GRCh38.dna.chromosome<br>export IDXDIR=/opt/human_grch38/hisat2_idx<br><br>export FILELIST=$FADIR/$&#123;FANAME&#125;.1.fa,$FADIR/$&#123;FANAME&#125;.2.fa,$FADIR/$&#123;FANAME&#125;.3.fa,$FADIR/$&#123;FANAME&#125;.4.fa,$FADIR/$&#123;FANAME&#125;.5.fa,$FADIR/$&#123;FANAME&#125;.6.fa,$FADIR/$&#123;FANAME&#125;.7.fa,$FADIR/$&#123;FANAME&#125;.8.fa,$FADIR/$&#123;FANAME&#125;.9.fa,$FADIR/$&#123;FANAME&#125;.10.fa,<br>export FILELIST=$&#123;FILELIST&#125;$FADIR/$&#123;FANAME&#125;.11.fa,$FADIR/$&#123;FANAME&#125;.12.fa,$FADIR/$&#123;FANAME&#125;.13.fa,$FADIR/$&#123;FANAME&#125;.14.fa,$FADIR/$&#123;FANAME&#125;.15.fa,$FADIR/$&#123;FANAME&#125;.16.fa,$FADIR/$&#123;FANAME&#125;.17.fa,$FADIR/$&#123;FANAME&#125;.18.fa,$FADIR/$&#123;FANAME&#125;.19.fa,$FADIR/$&#123;FANAME&#125;.20.fa,<br>export FILELIST=$&#123;FILELIST&#125;$FADIR/$&#123;FANAME&#125;.21.fa,$FADIR/$&#123;FANAME&#125;.22.fa,$FADIR/$&#123;FANAME&#125;.MT.fa,$FADIR/$&#123;FANAME&#125;.X.fa,$FADIR/$&#123;FANAME&#125;.Y.fa<br>echo **************************************<br>echo $FILELIST<br>echo **************************************<br>hisat2-build -p 8 $FILELIST $IDXDIR/GRCh38.hisat2.idx<br></code></pre></td></tr></table></figure><h2 id="Sam-Bam-bam-记录比对的具体情况">Sam/Bam(.bam) 记录比对的具体情况</h2><p><a href="https://hexo.limour.top/go/#aHR0cHM6Ly93d3cuamlhbnNodS5jb20vcC9mZjYxODdjOTcxNTU=" rel="noopener external nofollow noreferrer">Sam/Bam文件格式详解</a></p><p>bam文件是sam文件的二进制格式，sam 文件是Sequence Alignment/Map Format的简写，产生于比对之后的数据输出，记录了比对的具体情况。文件中以tab键分割，包括 <code>Header section</code> 和 <code>Alignments section</code> 两部分：</p><h3 id="Header-section">Header section</h3><p>该部分全部以“@”开头，提供基本的软件版本，参考序列信息，排序信息等</p><ul><li>@HD行：这一行中有各种不同的标识<ul><li>标识“VN”用以说明格式版本</li><li>标识“SO”用以说明比对排序的情况，有unknown (default)、unsorted、queryname和coordinate,对于coordinate，排序的主键是Alignments section的第三列“RNAME”，其顺序由@SQ行的“SN”标识的顺序定义，次要排序键是Alignments section的第四列“POS”字段。对于RNAME和POS相等的比对，排列顺序则是任意的</li></ul></li><li>@SQ行的“SN”标签是参考序列说明，它的值主要是用于Alignments section的第三列“RNAME”和第七列“MRNM”比对的记录</li><li>@PG行是使用的程序说明；该行“ID”为程序记录标识符，“PN”为程序名字，“CL”为命令行</li><li>@CO行是任意的说明信息</li></ul><h3 id="Alignments-section">Alignments section</h3><p>该部分包含了11列必需字段，无效或者没有的字段一般用<code>0</code>或者<code>*</code>表示。</p><figure class="highlight txt"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs txt">@HD VN:1.6 SO:coordinate<br>@SQ SN:ref LN:45<br>r001 99 ref 7 30 8M2I4M1D3M = 37 39 TTAGATAAAGGATACTG *<br>r002 0 ref 9 30 3S6M1P1I4M * 0 0 AAAAGATAAGGATA *<br>r003 0 ref 9 30 5S6M * 0 0 GCCTAAGCTAA * SA:Z:ref,29,-,6H5M,17,0;<br>r004 0 ref 16 30 6M14N5M * 0 0 ATAGCTTCAGC *<br>r003 2064 ref 29 17 6H5M * 0 0 TAGGC * SA:Z:ref,9,+,5S6M,30,1;<br>r001 147 ref 37 30 9M = 7 -39 CAGCGGCAT * NM:i:1<br></code></pre></td></tr></table></figure><h4 id="第1列：Qname">第1列：Qname</h4><p>Read的名字</p><h4 id="第2列：FLAG">第2列：FLAG</h4><p>每一个read的比对情况可以用十进制数字（或者十六进制数字）表示，如果比对情况 有多个，将多个比对情况所代表的十进制数字加和就是这一行的FLAG。<br>另，以下网站可以通过输入FLAG值，直接找出该FLAG是那些FLAG的加和：<a href="https://hexo.limour.top/go/#aHR0cHM6Ly9icm9hZGluc3RpdHV0ZS5naXRodWIuaW8vcGljYXJkL2V4cGxhaW4tZmxhZ3MuaHRtbA==" rel="noopener external nofollow noreferrer">Decoding SAM flags</a></p><h4 id="第3列：RNAME">第3列：RNAME</h4><p>比对上的参考序列的名字，该名字出现在Header section的@SQ行的SN标识中，如果该read没有比对上，也就是说该read在参考序列上没有坐标，那么这一列则用“”表示，那么这一行的POS和CIGAR列也会是“”。</p><h4 id="第4列：POS">第4列：POS</h4><p>read比对到的参考序列“RNAME”最左侧的位置坐标，也是CIGAR中第一个比对标识“M”对应的最左侧碱基在参考序列的位置，未比对上的read在参考序列中没有坐标，此列标识为“0”。</p><h4 id="第5列：MAPQ">第5列：MAPQ</h4><p>比对的质量值，计算方法为比对错误率的-10*log10的值，一般是四舍五入的整数值，如果是255，说明该比对值无效。</p><h4 id="第6列：CIGAR">第6列：CIGAR</h4><p>CIGAR标识符表示read中每个碱基的比对情况，主要有以下标识符：</p><ul><li>M: read上的碱基与参考序列“RNAME”完全匹配，碱基一一对应，包括了正确匹配与错误匹配</li><li>I: read上的碱基相对于参考序列“RNAME”有插入现象</li><li>D: read上的碱基相对于参考序列“RNAME”有删除现象</li><li>N: read上的碱基相对于参考序列“RNAME”存在连续没有比对上的空缺</li><li>S: read的开头或者结尾部分没有比对到参考序列&quot;RNAME”上, 但这部分未比对上的连续序列仍保留在sam文件的该read序列中</li><li>H: read的开头或者结尾部分没有比对到参考序列&quot;RNAME”上, 这部分未比对上的连续序列未保留</li><li>P: padding (silent deletion from padded reference)</li><li>=：sequence match 正确匹配</li><li>X：sequence mismatch 错误匹配</li></ul><h4 id="第7列：MRNM">第7列：MRNM</h4><p>该read的mate read比对上的参考序列的名字，该名字出现在Header section的@SQ行的SN标识中，</p><ul><li>如果和该read所在行的第三列“RNAME”一样，则用“=”表示，说明这对read比对到了同一条参考序列上；</li><li>如果mate read没有比对上，第七列则用“*”表示；</li><li>如果这对read没有比对到同一条参考序列，那么这一列则是mate read所在行第三列的“RNAME”。</li></ul><h4 id="第8列：MPOS">第8列：MPOS</h4><p>该read的mate read比对到的参考序列“RNAME”最左侧的位置坐标，也是mate read CIGAR中第一个比对标识“M”对应的最左侧碱基在参考序列的位置，未比对上的read在参考序列中没有坐标，此列标识为“0”。</p><h4 id="第9列：ISIZE">第9列：ISIZE</h4><p>表示pair read完全匹配到同一条参考序列时，两个read之间的长度，可简单理解为测序文库的长度。</p><h4 id="第10列：SEQ">第10列：SEQ</h4><p>存储的序列，没有存储，此列则用“*”标识。该序列的长度一定等于CIGAR标识中“M”，“I”，“S”，“=”，“X”标识的碱基长度之和。</p><h4 id="第11列：QUAL">第11列：QUAL</h4><p>序列的每个碱基对应一个碱基质量字符，每个碱基质量字符对应的ASCII码值减去33（Sanger Phred-33 质量值体系），即为该碱基的测序质量得分（Phred Quality Score）。不同Phred Quality Score代表不同的碱基测序错误率，如Phred Quality Score值为20和30分别表示碱基测序错误率为1%和0.1%。</p><h2 id="相关参数说明">相关参数说明</h2><p><a href="https://hexo.limour.top/go/#aHR0cHM6Ly93d3cuamllYW5kemUxMzE0LmNvbS9wb3N0L2NucG9zdHMvMTgv" rel="noopener external nofollow noreferrer">基因组的那些事儿</a></p><ul><li>测序深度：30x；每个碱基平均被测次数，相关研究表明5~60x中 30x对于后续分析可以达95%置信度</li><li>测序策略：PE150；PE双端测序、一条序列正反测两次；150每次测150bp，双端测一条片段共300bp</li><li>350bpcDNA建库：将DNA用超声波随机打断成350bp，加接头，作为测序前的准备工作</li></ul></div><hr><div><div class="post-metas my-3"><div class="post-meta"><i class="iconfont icon-tags"></i> <a href="/tags/ngs/" class="print-no-link">#NGS</a> <a href="/tags/fasta/" class="print-no-link">#fasta</a> <a href="/tags/fastq/" class="print-no-link">#fastq</a> <a href="/tags/gft/" class="print-no-link">#gft</a> <a href="/tags/bam/" class="print-no-link">#bam</a></div></div><div class="license-box my-3"><div class="license-title"><div>【迁移】二代测序数据处理之数据格式说明</div><div>https://hexo.limour.top/er-dai-ce-xu-shu-ju-chu-li-zhi-shu-ju-ge-shi-shuo-ming</div></div><div class="license-meta"><div class="license-meta-item"><div>Author</div><div>Limour</div></div><div class="license-meta-item license-meta-date"><div>Posted on</div><div>January 17, 2022</div></div><div class="license-meta-item license-meta-date"><div>Updated on</div><div>March 19, 2024</div></div><div class="license-meta-item"><div>Licensed under</div><div><a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><span class="hint--top hint--rounded" aria-label="BY - Attribution"><i class="iconfont icon-cc-by"></i> </span></a><a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><span class="hint--top hint--rounded" aria-label="NC - Non-commercial"><i class="iconfont icon-cc-nc"></i> </span></a><a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><span class="hint--top hint--rounded" aria-label="SA - Share-alike"><i class="iconfont icon-cc-sa"></i></span></a></div></div></div><div class="license-icon iconfont"></div></div><div class="post-prevnext my-3"><article class="post-prev col-6"><a href="/-qian-yi-CellTypist-zhu-shi-mian-yi-xi-bao-ya-qun" title="【迁移】CellTypist 注释免疫细胞亚群"><i class="iconfont icon-arrowleft"></i> <span class="hidden-mobile">【迁移】CellTypist 注释免疫细胞亚群</span> <span class="visible-mobile">Previous</span></a></article><article class="post-next col-6"><a href="/install-common-packages-with-conda" title="【记录】通过conda安装一些常用包"><span class="hidden-mobile">【记录】通过conda安装一些常用包</span> <span class="visible-mobile">Next</span> <i class="iconfont icon-arrowright"></i></a></article></div></div><article id="comments" lazyload><div id="waline"></div><script type="text/javascript">Fluid.utils.loadComments("#waline",function(){Fluid.utils.createCssLink("/cdn/waline/waline.css"),Fluid.utils.createScript("/cdn/waline/waline.js",function(){var i=Object.assign({serverURL:"https://comments.limour.top",path:"window.location.pathname",meta:["nick","mail","link"],requiredMeta:["nick"],lang:"zh-CN",emoji:["https://jscdn.limour.top/gh/walinejs/emojis/weibo"],dark:'html[data-user-color-scheme="dark"]',wordLimit:0,pageSize:10},{el:"#waline",path:window.location.pathname});Waline.init(i),Fluid.utils.waitElementVisible("#waline .vcontent",()=>{var i="#waline .vcontent img:not(.vemoji)";Fluid.plugins.imageCaption(i),Fluid.plugins.fancyBox(i)})})})</script><noscript>Please enable JavaScript to view the comments</noscript></article></article></div></div></div><div class="side-col d-none d-lg-block col-lg-2"><aside class="sidebar" style="margin-left:-1rem"><div id="toc"><p class="toc-header"><i class="iconfont icon-list"></i> <span>Table of Contents</span></p><div class="toc-body" id="toc-body"></div></div></aside></div></div></div><a id="scroll-top-button" aria-label="TOP" href="#" role="button"><i class="iconfont icon-arrowup" aria-hidden="true"></i></a><div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable modal-lg" role="document"><div class="modal-content"><div class="modal-header text-center"><h4 class="modal-title w-100 font-weight-bold">Search</h4><button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button></div><div class="modal-body mx-3"><div class="md-form mb-5"><input type="text" id="local-search-input" class="form-control validate"> <label data-error="x" data-success="v" for="local-search-input">Keyword</label></div><div class="list-group" id="local-search-result"></div></div></div></div></div></main><footer><div class="footer-inner"><div class="footer-content"><a target="_blank" rel="nofollow noopener" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=43130202000203"><img src="https://img.limour.top/2023/08/27/64eadeb81d6a0.webp" srcset="https://img.limour.top/inload.svg" lazyload>湘公网安备43130202000203号 </a><a target="_blank" rel="nofollow noopener" href="https://beian.miit.gov.cn/">湘ICP备20008299号 </a><a target="_blank" rel="nofollow noopener" href="https://icp.gov.moe/?keyword=20210128">萌ICP备20210128号</a><br><a href="https://www.foreverblog.cn/" target="_blank"><img src="https://img.foreverblog.cn/logo_en_default.png" srcset="https://img.limour.top/inload.svg" lazyload alt="" style="width:auto;height:24px"></a><br><a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/limour-blog/limour-blog.github.io" target="_blank" rel="nofollow noopener"><span>SRC</span></a> <i class="iconfont icon-love"></i> <a href="https://web.archive.org/web/20231130095837/https://effectiveacceleration.tech/" target="_blank" rel="nofollow noopener"><span>e/Acc</span></a></div></div></footer><script src="/cdn/nprogress/nprogress.min.js"></script><link rel="stylesheet" href="/cdn/nprogress/nprogress.min.css"><script>NProgress.configure({showSpinner:!1,trickleSpeed:100}),NProgress.start(),window.addEventListener("load",function(){NProgress.done()})</script><script src="/cdn/jquery/jquery.min.js"></script><script src="/cdn/bootstrap/js/bootstrap.min.js"></script><script src="/js/events.js"></script><script src="/js/plugins.js"></script><script src="/js/img-lazyload.js"></script><script>Fluid.utils.createScript("/cdn/tocbot/tocbot.min.js",function(){var t,o=jQuery("#toc");0!==o.length&&window.tocbot&&(t=jQuery("#board-ctn").offset().top,window.tocbot.init(Object.assign({tocSelector:"#toc-body",contentSelector:".markdown-body",linkClass:"tocbot-link",activeLinkClass:"tocbot-active-link",listClass:"tocbot-list",isCollapsedClass:"tocbot-is-collapsed",collapsibleClass:"tocbot-is-collapsible",scrollSmooth:!0,includeTitleTags:!0,headingsOffset:-t},CONFIG.toc)),0<o.find(".toc-list-item").length&&o.css("visibility","visible"),Fluid.events.registerRefreshCallback(function(){var t;"tocbot"in window&&(tocbot.refresh(),0!==(t=jQuery("#toc")).length)&&tocbot&&0<t.find(".toc-list-item").length&&t.css("visibility","visible")}))})</script><script src="/cdn/clipboard-js/clipboard.min.js"></script><script>Fluid.plugins.codeWidget()</script><script>Fluid.utils.createScript("/cdn/anchor-js/anchor.min.js",function(){window.anchors.options={placement:CONFIG.anchorjs.placement,visible:CONFIG.anchorjs.visible},CONFIG.anchorjs.icon&&(window.anchors.options.icon=CONFIG.anchorjs.icon);var n,o=[];for(n of(CONFIG.anchorjs.element||"h1,h2,h3,h4,h5,h6").split(","))o.push(".markdown-body > "+n.trim());"left"===CONFIG.anchorjs.placement&&(window.anchors.options.class="anchorjs-link-left"),window.anchors.add(o.join(", ")),Fluid.events.registerRefreshCallback(function(){if("anchors"in window){anchors.removeAll();var n,o=[];for(n of(CONFIG.anchorjs.element||"h1,h2,h3,h4,h5,h6").split(","))o.push(".markdown-body > "+n.trim());"left"===CONFIG.anchorjs.placement&&(anchors.options.class="anchorjs-link-left"),anchors.add(o.join(", "))}})})</script><script>Fluid.plugins.imageCaption()</script><script src="/js/local-search.js"></script><script src="/js/boot.js"></script><noscript><div class="noscript-warning">Blog works best with JavaScript enabled</div></noscript><script defer src="/theme-inject/timeliness.js"></script></body></html>