<!DOCTYPE html><html lang="en" data-default-color-scheme="auto"><head><script>let sw=navigator.serviceWorker;sw.ready.then(()=>{sw.controller||location.reload()}),sw.register("/sw.js",{scope:"/"})</script><script src="/theme-inject/cdn.js"></script><script defer src="https://api.limour.top/vue/0d2f95c1-755d-436b-adf8-eee12a80ed32/script.js"></script><meta charset="UTF-8"><link rel="apple-touch-icon" sizes="76x76" href="https://img.limour.top/2023/08/29/64ee07361815a.webp"><link rel="icon" href="https://img.limour.top/2023/08/29/64ee07361815a.webp"><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=5,shrink-to-fit=no"><meta http-equiv="x-ua-compatible" content="ie=edge"><meta name="theme-color" content="#2f4154"><meta name="author" content="Limour"><meta name="keywords" content=""><meta name="description" content="2023-09-27更新：添加了转录本计数相关内容            一些依赖 安装 biobambam  项目地址：https://gitlab.com/german.tischler/biobambam2  1conda create -n biobambam -c bioconda biobambam -y 安装 star 1234conda create -"><title>【迁移】STAR：一键脚本 - Limour&#39;s Blog</title><link rel="stylesheet" href="/cdn/bootstrap/css/bootstrap.min.css"><link rel="stylesheet" href="/cdn/github-markdown-css/github-markdown.min.css"><link rel="stylesheet" href="/cdn/hint.css/hint.min.css"><link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css"><link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_lbnruvf0jn.css"><link rel="stylesheet" href="/css/main.css"><link id="highlight-css" rel="stylesheet" href="/css/highlight.css"><link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css"><link rel="stylesheet" href="/theme-inject/custom.css"><link rel="stylesheet" href="/theme-inject/iconfont.css"><script id="fluid-configs">var dntVal,Fluid=window.Fluid||{},CONFIG=(Fluid.ctx=Object.assign({},Fluid.ctx),{hostname:"hexo.limour.top",root:"/",version:"1.9.8",typing:{enable:!1,typeSpeed:70,cursorChar:"_",loop:!1,scope:[]},anchorjs:{enable:!0,element:"h1,h2,h3,h4,h5,h6",placement:"left",visible:"hover",icon:"§"},progressbar:{enable:!0,height_px:3,color:"#29d",options:{showSpinner:!1,trickleSpeed:100}},code_language:{enable:!0,default:"TEXT"},copy_btn:!0,image_caption:{enable:!0},image_zoom:{enable:!1,img_url_replace:["",""]},toc:{enable:!0,placement:"right",headingSelector:"h1,h2,h3,h4,h5,h6",collapseDepth:0},lazyload:{enable:!0,loading_img:"https://img.limour.top/inload.svg",onlypost:!1,offset_factor:.5},web_analytics:{enable:!1,follow_dnt:!0,baidu:null,google:{measurement_id:null},tencent:{sid:null,cid:null},leancloud:{app_id:null,app_key:null,server_url:null,path:"window.location.pathname",ignore_local:!1},umami:{src:null,website_id:null,domains:null,start_time:"2024-01-01T00:00:00.000Z",token:null,api_server:null},woyaola:null,cnzz:null},search_path:"/local-search.xml",include_content_in_search:!0});CONFIG.web_analytics.follow_dnt&&(dntVal=navigator.doNotTrack||window.doNotTrack||navigator.msDoNotTrack,Fluid.ctx.dnt=dntVal&&(dntVal.startsWith("1")||dntVal.startsWith("yes")||dntVal.startsWith("on")))</script><script src="/js/utils.js"></script><script src="/js/color-schema.js"></script><link rel="canonical" href="https://hexo.limour.top/STAR--yi-jian-jiao-ben"><meta name="generator" content="Hexo 7.1.1"><link rel="alternate" href="/atom.xml" title="Limour's Blog" type="application/atom+xml"><link rel="alternate" href="/rss2.xml" title="Limour's Blog" type="application/rss+xml"></head><body><header><div class="header-inner" style="height:70vh"><nav id="navbar" class="navbar fixed-top navbar-expand-lg navbar-dark scrolling-navbar"><div class="container"><a class="navbar-brand" href="/"><strong>Limour&#39;s Blog</strong> </a><button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation"><div class="animated-icon"><span></span><span></span><span></span></div></button><div class="collapse navbar-collapse" id="navbarSupportedContent"><ul class="navbar-nav ml-auto text-center"><li class="nav-item"><a class="nav-link" href="https://hexo.limour.top/" target="_self"><i class="iconfont icon-home-fill"></i> <span>Home</span></a></li><li class="nav-item"><a class="nav-link" href="/archives/" target="_self"><i class="iconfont icon-archive-fill"></i> <span>Archive1</span></a></li><li class="nav-item"><a class="nav-link" href="https://occdn.limour.top/archives/" target="_self"><i class="iconfont icon-archive-fill"></i> <span>Archive2</span></a></li><li class="nav-item"><a class="nav-link" href="https://b.limour.top/archives/" target="_self"><i class="iconfont icon-archive-fill"></i> <span>Archive3</span></a></li><li class="nav-item"><a class="nav-link" href="https://od.limour.top/" target="_self"><i class="iconfont icon-onedrive"></i> <span>Alist</span></a></li><li class="nav-item"><a class="nav-link" href="https://orcid.org/0000-0001-8897-1685" target="_self"><i class="iconfont icon-orcid"></i> <span>Orcid</span></a></li><li class="nav-item"><a class="nav-link" href="/links/" target="_self"><i class="iconfont icon-link-fill"></i> <span>Links</span></a></li><li class="nav-item"><a class="nav-link" href="/atom.xml" target="_self"><i class="iconfont icon-rss"></i> <span>RSS</span></a></li><li class="nav-item" id="search-btn"><a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search"><i class="iconfont icon-search"></i></a></li><li class="nav-item" id="color-toggle-btn"><a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle"><i class="iconfont icon-dark" id="color-toggle-icon"></i></a></li></ul></div></div></nav><div id="banner" class="banner" parallax="true" style="background:url(https://img.limour.top/2023/08/29/64ee08e108638.webp) no-repeat center center;background-size:cover"><div class="full-bg-img"><div class="mask flex-center" style="background-color:rgba(0,0,0,.3)"><div class="banner-text text-center fade-in-up"><div class="h2"><span id="subtitle">【迁移】STAR：一键脚本</span></div><div class="mt-3"><span class="post-meta mr-2"><i class="iconfont icon-author" aria-hidden="true"></i> Limour </span><span class="post-meta"><i class="iconfont icon-date-fill" aria-hidden="true"></i> <time datetime="2022-07-02 18:17" pubdate>July 2, 2022 pm</time></span></div><div class="mt-1"><span class="post-meta mr-2"><i class="iconfont icon-chart"></i> 1.9k words </span><span class="post-meta mr-2"><i class="iconfont icon-clock-fill"></i> 16 mins</span></div></div></div></div></div></div></header><main><div class="container-fluid nopadding-x"><div class="row nomargin-x"><div class="side-col d-none d-lg-block col-lg-2"></div><div class="col-lg-8 nopadding-x-md"><div class="container nopadding-x-md" id="board-ctn"><div id="board"><article class="post-content mx-auto"><h1 id="seo-header">【迁移】STAR：一键脚本</h1><p id="updated-time" class="note note-info">Last updated on March 19, 2024 pm</p><div class="markdown-body"><p><img src="https://img.limour.top/2023/09/01/64f1bbec88640.webp" srcset="https://img.limour.top/inload.svg" lazyload alt="TCGA的STAR流程"></p><div class="note note-info"><p>2023-09-27更新：添加了转录本计数相关内容</p></div><h2 id="一些依赖">一些依赖</h2><h3 id="安装-biobambam">安装 biobambam</h3><ul><li>项目地址：<a href="https://hexo.limour.top/go/#aHR0cHM6Ly9naXRsYWIuY29tL2dlcm1hbi50aXNjaGxlci9iaW9iYW1iYW0y" rel="noopener external nofollow noreferrer">https://gitlab.com/german.tischler/biobambam2</a></li></ul><figure class="highlight mipsasm"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs mipsasm">conda create -n <span class="hljs-keyword">biobambam </span>-c <span class="hljs-keyword">bioconda </span><span class="hljs-keyword">biobambam </span>-y<br></code></pre></td></tr></table></figure><h3 id="安装-star">安装 star</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">conda create -n star -c bioconda star -y<br>conda activate star<br>conda install -c bioconda samtools -y<br>conda install -c bioconda rsem -y<br></code></pre></td></tr></table></figure><h2 id="可选：BAM转FASTQ">可选：BAM转FASTQ</h2><p>将所有bam文件以.bam结尾，单独存放到一个目录，conda activate biobambam，新建以下脚本运行</p><figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-meta">#!/bin/sh</span><br>ROOT=/home/jovyan/upload/22.07.02/muscle<br><span class="hljs-built_in">mkdir</span> <span class="hljs-variable">$ROOT</span><br> <br><span class="hljs-keyword">for</span> _ <span class="hljs-keyword">in</span> *.bam<br><span class="hljs-keyword">do</span><br> <br><span class="hljs-built_in">mkdir</span> <span class="hljs-variable">$ROOT</span>/<span class="hljs-variable">$&#123;_%.bam&#125;</span><br> <br>bamtofastq \<br>collate=1 \<br>exclude=QCFAIL,SECONDARY,SUPPLEMENTARY \<br>filename=<span class="hljs-variable">$_</span> \<br>gz=1 \<br>inputformat=bam \<br>level=5 \<br>outputdir=<span class="hljs-variable">$ROOT</span>/<span class="hljs-variable">$&#123;_%.bam&#125;</span> \<br>outputperreadgroup=1 \<br>outputperreadgroupsuffixF=_1.fq.gz \<br>outputperreadgroupsuffixF2=_2.fq.gz \<br>outputperreadgroupsuffixO=_o1.fq.gz \<br>outputperreadgroupsuffixO2=_o2.fq.gz \<br>outputperreadgroupsuffixS=_s.fq.gz \<br>tryoq=1 \<br> <br><span class="hljs-keyword">done</span><br></code></pre></td></tr></table></figure><ul><li>biobambam 高版本可能有 libmaus2 相关错误尚未修复</li><li>filename、outputdir 等参数等于号后不能有空格</li><li>单端测序，outputdir 里只有 default_s.fq.gz 的输出</li><li><a href="https://hexo.limour.top/go/#aHR0cHM6Ly9kb2NzLmdkYy5jYW5jZXIuZ292L0RhdGEvQmlvaW5mb3JtYXRpY3NfUGlwZWxpbmVzL0ROQV9TZXFfVmFyaWFudF9DYWxsaW5nX1BpcGVsaW5lLyNzdGVwLTEtY29udmVydGluZy1iYW1zLXRvLWZhc3Rxcy13aXRoLWJpb2JhbWJhbS1iaW9iYW1iYW0yLTIwNTQ=" rel="noopener external nofollow noreferrer">更多信息</a></li></ul><h2 id="首次：构建小鼠的基因组索引">首次：构建小鼠的基因组索引</h2><h3 id="来源">来源</h3><ul><li>human 和 mouse 推荐从 <a href="https://hexo.limour.top/go/#aHR0cHM6Ly93d3cuZ2VuY29kZWdlbmVzLm9yZy9tb3VzZQ==" rel="noopener external nofollow noreferrer">Gencode</a> 上下载</li><li>其他物种可以从 <a href="https://hexo.limour.top/go/#aHR0cHM6Ly93d3cubmNiaS5ubG0ubmloLmdvdi9nZW5vbWUvYnJvd3NlIyEvb3ZlcnZpZXc=" rel="noopener external nofollow noreferrer">NCBI</a> 上下载，也可以从 <a href="https://hexo.limour.top/go/#aHR0cHM6Ly93d3cuZW5zZW1ibC5vcmcvaW5mby9hYm91dC9zcGVjaWVzLmh0bWw=" rel="noopener external nofollow noreferrer">ENSEMBL</a> 上下载</li><li>需要参考基因组 fasta 文件和对应的 gtf 注释</li></ul><h3 id="步骤">步骤</h3><figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><code class="hljs sh">zcat XL789vs123/XL1/default_s.fq.gz | <span class="hljs-built_in">head</span> <span class="hljs-comment">#这个例子只有50, 因此 sjdbOverhang 为49</span><br> <br>wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M29/GRCm39.primary_assembly.genome.fa.gz<br>wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M29/gencode.vM29.primary_assembly.annotation.gtf.gz<br> <br>gunzip *.gz<br> <br>conda activate star<br>nano 2.sh<br><span class="hljs-built_in">chmod</span> +x 2.sh<br> <br><span class="hljs-comment">#!/bin/sh</span><br>STAR \<br>--runMode genomeGenerate \<br>--genomeDir index \<br>--genomeFastaFiles GRCm39.primary_assembly.genome.fa \<br>--sjdbOverhang 49 \<br>--sjdbGTFfile gencode.vM29.primary_assembly.annotation.gtf \<br>--runThreadN 8<br></code></pre></td></tr></table></figure><ul><li>如果要自己构建，可以使用 zcat R1.fq.gz | head 来查看reads长度，选用reads长度减1（149）作为 --sjdbOverhang 比默认的100要好，但是说明里认为绝大多数情况下100和理想值差不多</li></ul><h3 id="已构建好的索引">已构建好的索引</h3><ul><li><a href="https://hexo.limour.top/go/#aHR0cHM6Ly9nZGMuY2FuY2VyLmdvdi9hYm91dC1kYXRhL2dkYy1kYXRhLXByb2Nlc3NpbmcvZ2RjLXJlZmVyZW5jZS1maWxlcw==" rel="noopener external nofollow noreferrer">TCGA</a> 上有已经构建好的索引，但是可能是损坏的</li><li><code>wget https://api.gdc.cancer.gov/data/07f2dca9-cd39-4cbf-90d2-c7a1b8df5139 -O star-2.7.5c_GRCh38.d1.vd1_gencode.v36.tgz</code></li><li>tar -zxvf star-2.7.5c_GRCh38.d1.vd1_gencode.v36.tgz</li></ul><h3 id="附加-构建转录本索引">附加 构建转录本索引</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">cd</span> /home/jovyan/upload/zl_liu/star/ <span class="hljs-comment"># gtf文件和genome.fa文件所在目录</span><br>rsem-prepare-reference --gtf gencode.v36.primary_assembly.annotation.gtf GRCh38.primary_assembly.genome.fa GRCh38.primary_assembly.rsem.idx -p 8<br>grep -e <span class="hljs-string">&#x27;^ENST&#x27;</span> GRCh38.primary_assembly.rsem.idx.ti &gt; GRCh38.primary_assembly.rsem.t2n.ti<br>grep -e <span class="hljs-string">&#x27;^ENSG&#x27;</span> GRCh38.primary_assembly.rsem.idx.ti &gt; GRCh38.primary_assembly.rsem.g2n.ti<br></code></pre></td></tr></table></figure><ul><li>大概生成了以下文件</li></ul><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs bash">├── [5.8M]  ./GRCh38.primary_assembly.rsem.g2n.ti<br>├── [ 770]  ./GRCh38.primary_assembly.rsem.idx.chrlist<br>├── [404K]  ./GRCh38.primary_assembly.rsem.idx.grp<br>├── [362M]  ./GRCh38.primary_assembly.rsem.idx.idx.fa<br>├── [362M]  ./GRCh38.primary_assembly.rsem.idx.n2g.idx.fa<br>├── [386M]  ./GRCh38.primary_assembly.rsem.idx.seq<br>├── [128M]  ./GRCh38.primary_assembly.rsem.idx.ti<br>├── [362M]  ./GRCh38.primary_assembly.rsem.idx.transcripts.fa<br>├── [6.5M]  ./GRCh38.primary_assembly.rsem.t2n.ti<br></code></pre></td></tr></table></figure><h2 id="一键脚本">一键脚本</h2><ul><li><a href="/shi-yong-GATK-zhao-SNP#fastp%E4%B8%80%E9%94%AE%E8%B4%A8%E6%8E%A7">原始数据质控</a></li></ul><p>单独起一个目录，在这个目录下，以样本名为目录名，将样本的fataq文件以gzip压缩后存放，单端一个文件，双端两个文件，都可以，示例结构如下</p><figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><code class="hljs sh">XL1_12/<br>├── XL01<br>│   └── default_s.fq.gz<br>├── XL02<br>│   └── default_s.fq.gz<br>├── XL03<br>│   └── default_s.fq.gz<br>├── XL04<br>│   └── default_s.fq.gz<br>├── XL05<br>│   └── default_s.fq.gz<br>├── XL06<br>│   └── default_s.fq.gz<br>├── XL07<br>│   └── default_s.fq.gz<br>├── XL08<br>│   └── default_s.fq.gz<br>├── XL09<br>│   └── default_s.fq.gz<br>├── XL10<br>│   └── default_s.fq.gz<br>├── XL11<br>│   └── default_s.fq.gz<br>└── XL12<br>    └── default_s.fq.gz<br></code></pre></td></tr></table></figure><p>在这个目录外，conda activate star，新建以下脚本运行，即可跑完前五步</p><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-meta">#!/bin/bash</span><br><span class="hljs-built_in">source</span> activate star<br><span class="hljs-comment">#任务名</span><br>TASKN=XL1_12<br><span class="hljs-comment">#设置CleanData存放目录</span><br>CLEAN=/home/jovyan/upload/22.07.02/<span class="hljs-variable">$TASKN</span><br><span class="hljs-comment">#设置输出目录</span><br>WORK=/home/jovyan/upload/22.07.02/output_<span class="hljs-variable">$TASKN</span><br><span class="hljs-comment">#设置index目录</span><br>INDEX=/home/jovyan/upload/22.07.02/index<br><span class="hljs-comment">#设置参考文件位置</span><br>Reference=/home/jovyan/upload/22.07.02/GRCm39.primary_assembly.genome.fa<br><span class="hljs-comment">#设置 sjdbOverhang</span><br>sjdbOverhang=49<br><span class="hljs-comment">#设置 IIG 目录(这一步的输出目录)</span><br>IIG=/home/jovyan/upload/22.07.02/IIG_<span class="hljs-variable">$TASKN</span><br> <br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CLEAN</span><span class="hljs-string">&quot;, &quot;</span><span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;, &quot;</span><span class="hljs-variable">$INDEX</span><br><span class="hljs-built_in">mkdir</span> <span class="hljs-variable">$WORK</span><br> <br>CDIR=$(<span class="hljs-built_in">basename</span> `<span class="hljs-built_in">pwd</span>`)<br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CDIR</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CLEAN</span><br><span class="hljs-keyword">for</span> file <span class="hljs-keyword">in</span> <span class="hljs-variable">$CLEAN</span>/*<br><span class="hljs-keyword">do</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$file</span><br>SAMPLE=<span class="hljs-variable">$&#123;file##*/&#125;</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$SAMPLE</span><br><span class="hljs-built_in">mkdir</span> <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><br><span class="hljs-built_in">cd</span> <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><br> <br>STAR \<br>--genomeDir <span class="hljs-variable">$INDEX</span> \<br>--readFilesIn `<span class="hljs-built_in">ls</span> <span class="hljs-variable">$CLEAN</span>/<span class="hljs-variable">$SAMPLE</span>/*` \<br>--runThreadN 4 \<br>--outFilterMultimapScoreRange 1 \<br>--outFilterMultimapNmax 20 \<br>--outFilterMismatchNmax 10 \<br>--alignIntronMax 500000 \<br>--alignMatesGapMax 1000000 \<br>--sjdbScore 2 \<br>--alignSJDBoverhangMin 1 \<br>--genomeLoad LoadAndRemove \<br>--readFilesCommand zcat \<br>--outFilterMatchNminOverLread 0.33 \<br>--outFilterScoreMinOverLread 0.33 \<br>--sjdbOverhang <span class="hljs-variable">$sjdbOverhang</span> \<br>--outSAMstrandField intronMotif \<br>--outSAMtype None \<br>--outSAMmode None \<br> <br><span class="hljs-keyword">done</span><br> <br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CLEAN</span><span class="hljs-string">&quot;, &quot;</span><span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;, &quot;</span><span class="hljs-variable">$INDEX</span><span class="hljs-string">&quot;, &quot;</span><span class="hljs-variable">$IIG</span><br><span class="hljs-built_in">mkdir</span> <span class="hljs-variable">$IIG</span><br> <br>CDIR=$(<span class="hljs-built_in">basename</span> `<span class="hljs-built_in">pwd</span>`)<br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CDIR</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CLEAN</span><br><span class="hljs-keyword">for</span> file <span class="hljs-keyword">in</span> <span class="hljs-variable">$CLEAN</span>/*<br><span class="hljs-keyword">do</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$file</span><br>SAMPLE=<span class="hljs-variable">$&#123;file##*/&#125;</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><br><span class="hljs-keyword">done</span><br> <br>STAR \<br>--runMode genomeGenerate \<br>--genomeDir <span class="hljs-variable">$IIG</span> \<br>--genomeFastaFiles <span class="hljs-variable">$Reference</span> \<br>--sjdbOverhang <span class="hljs-variable">$sjdbOverhang</span> \<br>--runThreadN 4 \<br>--sjdbFileChrStartEnd `<span class="hljs-built_in">ls</span> <span class="hljs-variable">$WORK</span>/*/SJ.out.tab`<br> <br><span class="hljs-built_in">ln</span> -s <span class="hljs-variable">$INDEX</span>/exonGeTrInfo.tab <span class="hljs-variable">$IIG</span><br><span class="hljs-built_in">ln</span> -s <span class="hljs-variable">$INDEX</span>/exonInfo.tab <span class="hljs-variable">$IIG</span><br><span class="hljs-built_in">ln</span> -s <span class="hljs-variable">$INDEX</span>/geneInfo.tab <span class="hljs-variable">$IIG</span><br><span class="hljs-built_in">ln</span> -s <span class="hljs-variable">$INDEX</span>/sjdbList.fromGTF.out.tab <span class="hljs-variable">$IIG</span><br><span class="hljs-built_in">ln</span> -s <span class="hljs-variable">$INDEX</span>/transcriptInfo.tab <span class="hljs-variable">$IIG</span><br> <br>CDIR=$(<span class="hljs-built_in">basename</span> `<span class="hljs-built_in">pwd</span>`)<br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CDIR</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$CLEAN</span><br><span class="hljs-keyword">for</span> file <span class="hljs-keyword">in</span> <span class="hljs-variable">$CLEAN</span>/*<br><span class="hljs-keyword">do</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$file</span><br>SAMPLE=<span class="hljs-variable">$&#123;file##*/&#125;</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><br><span class="hljs-built_in">mkdir</span> <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><span class="hljs-string">&quot;/Res&quot;</span><br><span class="hljs-built_in">cd</span> <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><span class="hljs-string">&quot;/Res&quot;</span><br> <br>STAR \<br>--genomeDir <span class="hljs-variable">$IIG</span> \<br>--readFilesIn `<span class="hljs-built_in">ls</span> <span class="hljs-variable">$CLEAN</span>/<span class="hljs-variable">$SAMPLE</span>/*` \<br>--runThreadN 8 \<br>--quantMode TranscriptomeSAM GeneCounts \<br>--outFilterMultimapScoreRange 1 \<br>--outFilterMultimapNmax 20 \<br>--outFilterMismatchNmax 10 \<br>--alignIntronMax 500000 \<br>--alignMatesGapMax 1000000 \<br>--sjdbScore 2 \<br>--alignSJDBoverhangMin 1 \<br>--genomeLoad LoadAndRemove \<br>--limitBAMsortRAM  35000000000 \<br>--readFilesCommand zcat \<br>--outFilterMatchNminOverLread 0.33 \<br>--outFilterScoreMinOverLread 0.33 \<br>--sjdbOverhang <span class="hljs-variable">$sjdbOverhang</span> \<br>--outSAMstrandField intronMotif \<br>--outSAMattributes NH HI NM MD AS XS \<br>--outSAMunmapped Within \<br>--outSAMtype BAM SortedByCoordinate \<br>--outSAMheaderHD @HD VN:1.4 \<br>--outSAMattrRGline ID:sample SM:sample PL:ILLUMINA<br> <br><span class="hljs-keyword">done</span><br></code></pre></td></tr></table></figure><h2 id="附加-转录本计数">附加 转录本计数</h2><ul><li>转录本水平: cufflinks、eXpress、RSEM</li><li>外显子水平: DEXseq</li></ul><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs bash">nano RSEM.sh &amp;&amp; <span class="hljs-built_in">chmod</span> +x RSEM.sh<br>./RSEM.sh<br></code></pre></td></tr></table></figure><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-meta">#!/bin/bash</span><br><span class="hljs-built_in">source</span> activate star<br> <br><span class="hljs-comment">#任务名</span><br>TASKN=cleandata<br><span class="hljs-comment">#设置上一步的输出目录</span><br>WORK=/home/jovyan/work/shRNA-<span class="hljs-built_in">seq</span>/output_<span class="hljs-variable">$TASKN</span><br><span class="hljs-comment">#设置参考文件位置</span><br>Reference=/home/jovyan/upload/zl_liu/star/GRCh38.primary_assembly.genome.fa<br><span class="hljs-comment">#设置转录本索引位置(前缀)</span><br>RSEM_IDX=/home/jovyan/upload/zl_liu/star/GRCh38.primary_assembly.rsem.idx<br><span class="hljs-comment">#设置这一步的输出目录</span><br>RSEM_OUT=/home/jovyan/work/shRNA-<span class="hljs-built_in">seq</span>/RSEM_<span class="hljs-variable">$TASKN</span><br><span class="hljs-built_in">mkdir</span> -p <span class="hljs-variable">$RSEM_OUT</span><br> <br><span class="hljs-keyword">for</span> file <span class="hljs-keyword">in</span> <span class="hljs-variable">$WORK</span>/*<br><span class="hljs-keyword">do</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$file</span><br>SAMPLE=<span class="hljs-variable">$&#123;file##*/&#125;</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$SAMPLE</span><br><span class="hljs-built_in">mkdir</span> <span class="hljs-variable">$RSEM_OUT</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><br><span class="hljs-built_in">cd</span> <span class="hljs-variable">$RSEM_OUT</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><br> <br>rsem-calculate-expression -no-bam-output --alignments -p 8 --paired-end \<br>    <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><span class="hljs-string">&quot;/Res/Aligned.toTranscriptome.out.bam&quot;</span> \<br>    <span class="hljs-variable">$RSEM_IDX</span> \<br>    <span class="hljs-string">&#x27;resm&#x27;</span><br> <br><span class="hljs-keyword">done</span><br></code></pre></td></tr></table></figure><ul><li>最后得到的RSEM_OUT目录结构如下</li></ul><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><code class="hljs bash">[158M]  .<br>├── [ 26M]  ./NC-1<br>│   ├── [6.8M]  ./NC-1/resm.genes.results<br>│   ├── [ 15M]  ./NC-1/resm.isoforms.results<br>│   └── [4.3M]  ./NC-1/resm.stat<br>│       ├── [ 871]  ./NC-1/resm.stat/resm.cnt<br>│       ├── [493K]  ./NC-1/resm.stat/resm.model<br>│       └── [3.9M]  ./NC-1/resm.stat/resm.theta<br>├── [ 26M]  ./NC-2<br>│   ├── [6.8M]  ./NC-2/resm.genes.results<br>│   ├── [ 15M]  ./NC-2/resm.isoforms.results<br>│   └── [4.4M]  ./NC-2/resm.stat<br>│       ├── [ 872]  ./NC-2/resm.stat/resm.cnt<br>│       ├── [493K]  ./NC-2/resm.stat/resm.model<br>│       └── [3.9M]  ./NC-2/resm.stat/resm.theta<br>├── [ 26M]  ./NC-3<br>│   ├── [6.8M]  ./NC-3/resm.genes.results<br>│   ├── [ 15M]  ./NC-3/resm.isoforms.results<br>│   └── [4.4M]  ./NC-3/resm.stat<br>│       ├── [ 872]  ./NC-3/resm.stat/resm.cnt<br>│       ├── [493K]  ./NC-3/resm.stat/resm.model<br>│       └── [3.9M]  ./NC-3/resm.stat/resm.theta<br>├── [ 26M]  ./shRNA1-1<br>│   ├── [6.8M]  ./shRNA1-1/resm.genes.results<br>│   ├── [ 15M]  ./shRNA1-1/resm.isoforms.results<br>│   └── [4.4M]  ./shRNA1-1/resm.stat<br>│       ├── [ 874]  ./shRNA1-1/resm.stat/resm.cnt<br>│       ├── [493K]  ./shRNA1-1/resm.stat/resm.model<br>│       └── [3.9M]  ./shRNA1-1/resm.stat/resm.theta<br>├── [ 26M]  ./shRNA1-2<br>│   ├── [6.8M]  ./shRNA1-2/resm.genes.results<br>│   ├── [ 15M]  ./shRNA1-2/resm.isoforms.results<br>│   └── [4.3M]  ./shRNA1-2/resm.stat<br>│       ├── [ 849]  ./shRNA1-2/resm.stat/resm.cnt<br>│       ├── [493K]  ./shRNA1-2/resm.stat/resm.model<br>│       └── [3.8M]  ./shRNA1-2/resm.stat/resm.theta<br>└── [ 26M]  ./shRNA1-3<br>    ├── [6.8M]  ./shRNA1-3/resm.genes.results<br>    ├── [ 15M]  ./shRNA1-3/resm.isoforms.results<br>    └── [4.3M]  ./shRNA1-3/resm.stat<br>        ├── [ 867]  ./shRNA1-3/resm.stat/resm.cnt<br>        ├── [493K]  ./shRNA1-3/resm.stat/resm.model<br>        └── [3.8M]  ./shRNA1-3/resm.stat/resm.theta<br></code></pre></td></tr></table></figure><ul><li>组装RSEM_Counts文件</li></ul><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment">#设置转录本索引转换的位置(前缀)(无idx)</span><br>RSEM_x2n<span class="hljs-operator">=</span><span class="hljs-string">&#x27;/home/jovyan/upload/zl_liu/star/GRCh38.primary_assembly.rsem&#x27;</span><br><span class="hljs-comment">#设置RSEM的输出目录</span><br>RSEM_OUT<span class="hljs-operator">=</span><span class="hljs-string">&#x27;/home/jovyan/work/shRNA-seq/RSEM_cleandata&#x27;</span><br>file_list <span class="hljs-operator">&lt;-</span> list.dirs<span class="hljs-punctuation">(</span>RSEM_OUT<span class="hljs-punctuation">,</span> recursive<span class="hljs-operator">=</span><span class="hljs-built_in">F</span><span class="hljs-punctuation">,</span> full.names <span class="hljs-operator">=</span> <span class="hljs-built_in">F</span><span class="hljs-punctuation">)</span><br>t2n <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span>file <span class="hljs-operator">=</span> paste0<span class="hljs-punctuation">(</span>RSEM_x2n<span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;.t2n.ti&#x27;</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">,</span> sep <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;\t&#x27;</span><span class="hljs-punctuation">,</span> header <span class="hljs-operator">=</span> <span class="hljs-built_in">F</span><span class="hljs-punctuation">)</span><br>g2n <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span>file <span class="hljs-operator">=</span> paste0<span class="hljs-punctuation">(</span>RSEM_x2n<span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;.g2n.ti&#x27;</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">,</span> sep <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;\t&#x27;</span><span class="hljs-punctuation">,</span> header <span class="hljs-operator">=</span> <span class="hljs-built_in">F</span><span class="hljs-punctuation">)</span><br>transcript_counts <span class="hljs-operator">&lt;-</span> cbind<span class="hljs-punctuation">(</span>t2n<span class="hljs-punctuation">,</span> g2n<span class="hljs-punctuation">)</span><br>colnames<span class="hljs-punctuation">(</span>transcript_counts<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&#x27;transcript_id&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;transcript_name&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;gene_id&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;symbol&#x27;</span><span class="hljs-punctuation">)</span><br>transcript_tpm <span class="hljs-operator">&lt;-</span> transcript_counts<br>transcript_IsoPct <span class="hljs-operator">&lt;-</span> transcript_counts<br><span class="hljs-keyword">for</span><span class="hljs-punctuation">(</span>sample <span class="hljs-keyword">in</span> file_list<span class="hljs-punctuation">)</span><span class="hljs-punctuation">&#123;</span><br>    test_tab <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span>file <span class="hljs-operator">=</span> file.path<span class="hljs-punctuation">(</span>RSEM_OUT<span class="hljs-punctuation">,</span> sample<span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;resm.isoforms.results&#x27;</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">,</span> sep <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;\t&#x27;</span><span class="hljs-punctuation">,</span> header <span class="hljs-operator">=</span> <span class="hljs-built_in">T</span><span class="hljs-punctuation">)</span><br>    transcript_counts<span class="hljs-punctuation">[</span>sample<span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> test_tab<span class="hljs-operator">$</span>expected_count<br>    transcript_tpm<span class="hljs-punctuation">[</span>sample<span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> test_tab<span class="hljs-operator">$</span>TPM<br>    transcript_IsoPct<span class="hljs-punctuation">[</span>sample<span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> test_tab<span class="hljs-operator">$</span>IsoPct<br><span class="hljs-punctuation">&#125;</span><br>write.csv<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> transcript_counts<span class="hljs-punctuation">,</span> file <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;transcript_counts.csv&#x27;</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># 期望计数</span><br>write.csv<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> transcript_tpm<span class="hljs-punctuation">,</span> file <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;transcript_tpm.csv&#x27;</span><span class="hljs-punctuation">)</span><br>write.csv<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> transcript_IsoPct<span class="hljs-punctuation">,</span> file <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;transcript_IsoPct.csv&#x27;</span><span class="hljs-punctuation">)</span> <span class="hljs-comment"># 该转录本占所属基因的比例</span><br></code></pre></td></tr></table></figure><p><img src="https://img.limour.top/2023/09/27/6513f15871a29.webp" srcset="https://img.limour.top/inload.svg" lazyload alt="组装的转录本计数文件格式"></p><h2 id="组装Counts文件">组装Counts文件</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment">#设置第三步的输出目录</span><br>WORK<span class="hljs-operator">=</span><span class="hljs-string">&#x27;/home/jovyan/upload/22.07.02/output_XL789vs123&#x27;</span><br><span class="hljs-comment">#设置index中基因注释位置</span><br>Reference<span class="hljs-operator">=</span><span class="hljs-string">&#x27;/home/jovyan/upload/22.07.02/index/geneInfo.tab&#x27;</span><br>file_list <span class="hljs-operator">&lt;-</span> list.dirs<span class="hljs-punctuation">(</span>WORK<span class="hljs-punctuation">,</span> recursive<span class="hljs-operator">=</span><span class="hljs-built_in">F</span><span class="hljs-punctuation">,</span> full.names <span class="hljs-operator">=</span> <span class="hljs-built_in">F</span><span class="hljs-punctuation">)</span><br>geneN <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span>file <span class="hljs-operator">=</span> Reference<span class="hljs-punctuation">,</span> sep <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;\t&#x27;</span><span class="hljs-punctuation">,</span> skip <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-punctuation">)</span><br>colnames<span class="hljs-punctuation">(</span>geneN<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&#x27;ID&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;symbol&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;type&#x27;</span><span class="hljs-punctuation">)</span><br><span class="hljs-keyword">for</span><span class="hljs-punctuation">(</span>sample <span class="hljs-keyword">in</span> file_list<span class="hljs-punctuation">)</span><span class="hljs-punctuation">&#123;</span><br>    test_tab <span class="hljs-operator">&lt;-</span> read.table<span class="hljs-punctuation">(</span>file <span class="hljs-operator">=</span> file.path<span class="hljs-punctuation">(</span>WORK<span class="hljs-punctuation">,</span> sample<span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;Res&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;ReadsPerGene.out.tab&#x27;</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">,</span> sep <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;\t&#x27;</span><span class="hljs-punctuation">,</span> header <span class="hljs-operator">=</span> <span class="hljs-built_in">F</span><span class="hljs-punctuation">)</span><br>    test_tab <span class="hljs-operator">&lt;-</span> test_tab<span class="hljs-punctuation">[</span><span class="hljs-operator">-</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">4</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><br>    geneN<span class="hljs-punctuation">[</span>sample<span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> test_tab<span class="hljs-punctuation">[</span><span class="hljs-number">2</span><span class="hljs-punctuation">]</span><br><span class="hljs-punctuation">&#125;</span><br>write.csv<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> geneN<span class="hljs-punctuation">,</span> file <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;XL789vs123.csv&#x27;</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure><p><img src="https://img.limour.top/2023/09/06/64f7f7f39f19a.webp" srcset="https://img.limour.top/inload.svg" lazyload alt="组装的Counts文件格式"></p><h2 id="DESeq2分析">DESeq2分析</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><code class="hljs R">library<span class="hljs-punctuation">(</span>DESeq2<span class="hljs-punctuation">)</span><br>count_all <span class="hljs-operator">&lt;-</span> read.csv<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;liver.csv&quot;</span><span class="hljs-punctuation">,</span>header<span class="hljs-operator">=</span><span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span>row.names<span class="hljs-operator">=</span><span class="hljs-number">1</span><span class="hljs-punctuation">)</span><br>count_all<br>cts_b <span class="hljs-operator">&lt;-</span> count_all<span class="hljs-punctuation">[</span> <span class="hljs-punctuation">,</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-operator">-</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span><span class="hljs-operator">-</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span><span class="hljs-operator">-</span><span class="hljs-number">3</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">]</span><br>rownames<span class="hljs-punctuation">(</span>cts_b<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> count_all<span class="hljs-operator">$</span>ID<br>cts_bb <span class="hljs-operator">&lt;-</span> cts_b<br> <br>cts_b <span class="hljs-operator">&lt;-</span> cts_bb<span class="hljs-punctuation">[</span><span class="hljs-punctuation">,</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&#x27;XL07&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;XL08&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;XL09&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;XL10&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;XL11&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;XL12&#x27;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">]</span><br>keep <span class="hljs-operator">&lt;-</span> rowSums<span class="hljs-punctuation">(</span>cts_b<span class="hljs-punctuation">)</span> <span class="hljs-operator">&gt;</span> <span class="hljs-number">10</span><br>cts_b<span class="hljs-punctuation">[</span>keep<span class="hljs-punctuation">,</span><span class="hljs-punctuation">]</span><br>conditions <span class="hljs-operator">&lt;-</span> factor<span class="hljs-punctuation">(</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-built_in">rep</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;Control&quot;</span><span class="hljs-punctuation">,</span><span class="hljs-number">3</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">rep</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;XL&quot;</span><span class="hljs-punctuation">,</span><span class="hljs-number">3</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span>    <br>colData_b <span class="hljs-operator">&lt;-</span> data.frame<span class="hljs-punctuation">(</span>row.names <span class="hljs-operator">=</span> colnames<span class="hljs-punctuation">(</span>cts_b<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> conditions<span class="hljs-punctuation">)</span><br>colData_b<br>dds <span class="hljs-operator">&lt;-</span> DESeqDataSetFromMatrix<span class="hljs-punctuation">(</span>countData <span class="hljs-operator">=</span> cts_b<span class="hljs-punctuation">[</span>keep<span class="hljs-punctuation">,</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>                              colData <span class="hljs-operator">=</span> colData_b<span class="hljs-punctuation">,</span><br>                              design <span class="hljs-operator">=</span> <span class="hljs-operator">~</span> conditions<span class="hljs-punctuation">)</span><br>dds <span class="hljs-operator">&lt;-</span> DESeq<span class="hljs-punctuation">(</span>dds<span class="hljs-punctuation">)</span><br>res <span class="hljs-operator">&lt;-</span> results<span class="hljs-punctuation">(</span>dds<span class="hljs-punctuation">)</span><br>rres <span class="hljs-operator">&lt;-</span> cbind<span class="hljs-punctuation">(</span>count_all<span class="hljs-punctuation">[</span>keep<span class="hljs-punctuation">,</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span><span class="hljs-number">3</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span> data.frame<span class="hljs-punctuation">(</span>res<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>write.csv<span class="hljs-punctuation">(</span>rres<span class="hljs-punctuation">,</span> file<span class="hljs-operator">=</span><span class="hljs-string">&#x27;XL101112vs789_DESeq2.csv&#x27;</span><span class="hljs-punctuation">)</span><br>rres<br></code></pre></td></tr></table></figure><h2 id="附加：删除临时文件">附加：删除临时文件</h2><figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs sh"><span class="hljs-meta">#!/bin/sh</span><br><span class="hljs-comment">#设置CleanData存放目录</span><br>CLEAN=/home/jovyan/upload/yy_zhang（备份）/RNA-<span class="hljs-built_in">seq</span>/Cleandata<br><span class="hljs-comment">#设置这一步的输出目录 (确保目录存在)</span><br>WORK=/home/jovyan/upload/zl_liu/star_data/yyz_01/output<br> <br><span class="hljs-keyword">for</span> file <span class="hljs-keyword">in</span> <span class="hljs-variable">$CLEAN</span>/*<br><span class="hljs-keyword">do</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$file</span><br>SAMPLE=<span class="hljs-variable">$&#123;file##*/&#125;</span><br><span class="hljs-built_in">echo</span> <span class="hljs-variable">$SAMPLE</span><br> <br><span class="hljs-built_in">rm</span> -rf <span class="hljs-variable">$WORK</span><span class="hljs-string">&quot;/&quot;</span><span class="hljs-variable">$SAMPLE</span><span class="hljs-string">&quot;/IIG&quot;</span><br> <br><span class="hljs-keyword">done</span><br></code></pre></td></tr></table></figure></div><hr><div><div class="post-metas my-3"><div class="post-meta"><i class="iconfont icon-tags"></i> <a href="/tags/star/" class="print-no-link">#STAR</a></div></div><div class="license-box my-3"><div class="license-title"><div>【迁移】STAR：一键脚本</div><div>https://hexo.limour.top/STAR--yi-jian-jiao-ben</div></div><div class="license-meta"><div class="license-meta-item"><div>Author</div><div>Limour</div></div><div class="license-meta-item license-meta-date"><div>Posted on</div><div>July 2, 2022</div></div><div class="license-meta-item license-meta-date"><div>Updated on</div><div>March 19, 2024</div></div><div class="license-meta-item"><div>Licensed under</div><div><a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><span class="hint--top hint--rounded" aria-label="BY - Attribution"><i class="iconfont icon-cc-by"></i> </span></a><a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><span class="hint--top hint--rounded" aria-label="NC - Non-commercial"><i class="iconfont icon-cc-nc"></i> </span></a><a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><span class="hint--top hint--rounded" aria-label="SA - Share-alike"><i class="iconfont icon-cc-sa"></i></span></a></div></div></div><div class="license-icon iconfont"></div></div><div class="post-prevnext my-3"><article class="post-prev col-6"><a href="/-qian-yi--shi-kong-tu--yi-tian-lang-xing-dan-cheng-l-xing-wei-li" title="【迁移】时空图：以天狼星单程旅行为例"><i class="iconfont icon-arrowleft"></i> <span class="hidden-mobile">【迁移】时空图：以天狼星单程旅行为例</span> <span class="visible-mobile">Previous</span></a></article><article class="post-next col-6"><a href="/-qian-yi-CellTypist-zhu-shi-mian-yi-xi-bao-ya-qun" title="【迁移】CellTypist 注释免疫细胞亚群"><span class="hidden-mobile">【迁移】CellTypist 注释免疫细胞亚群</span> <span class="visible-mobile">Next</span> <i class="iconfont icon-arrowright"></i></a></article></div></div><article id="comments" lazyload><div id="waline"></div><script type="text/javascript">Fluid.utils.loadComments("#waline",function(){Fluid.utils.createCssLink("/cdn/waline/waline.css"),Fluid.utils.createScript("/cdn/waline/waline.js",function(){var i=Object.assign({serverURL:"https://comments.limour.top",path:"window.location.pathname",meta:["nick","mail","link"],requiredMeta:["nick"],lang:"zh-CN",emoji:["https://jscdn.limour.top/gh/walinejs/emojis/weibo"],dark:'html[data-user-color-scheme="dark"]',wordLimit:0,pageSize:10},{el:"#waline",path:window.location.pathname});Waline.init(i),Fluid.utils.waitElementVisible("#waline .vcontent",()=>{var i="#waline .vcontent img:not(.vemoji)";Fluid.plugins.imageCaption(i),Fluid.plugins.fancyBox(i)})})})</script><noscript>Please enable JavaScript to view the comments</noscript></article></article></div></div></div><div class="side-col d-none d-lg-block col-lg-2"><aside class="sidebar" style="margin-left:-1rem"><div id="toc"><p class="toc-header"><i class="iconfont icon-list"></i> <span>Table of Contents</span></p><div class="toc-body" id="toc-body"></div></div></aside></div></div></div><a id="scroll-top-button" aria-label="TOP" href="#" role="button"><i class="iconfont icon-arrowup" aria-hidden="true"></i></a><div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable modal-lg" role="document"><div class="modal-content"><div class="modal-header text-center"><h4 class="modal-title w-100 font-weight-bold">Search</h4><button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button></div><div class="modal-body mx-3"><div class="md-form mb-5"><input type="text" id="local-search-input" class="form-control validate"> <label data-error="x" data-success="v" for="local-search-input">Keyword</label></div><div class="list-group" id="local-search-result"></div></div></div></div></div></main><footer><div class="footer-inner"><div class="footer-content"><a target="_blank" rel="nofollow noopener" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=43130202000203"><img src="https://img.limour.top/2023/08/27/64eadeb81d6a0.webp" srcset="https://img.limour.top/inload.svg" lazyload>湘公网安备43130202000203号 </a><a target="_blank" rel="nofollow noopener" href="https://beian.miit.gov.cn/">湘ICP备20008299号 </a><a target="_blank" rel="nofollow noopener" href="https://icp.gov.moe/?keyword=20210128">萌ICP备20210128号</a><br><a href="https://www.foreverblog.cn/" target="_blank"><img src="https://img.foreverblog.cn/logo_en_default.png" srcset="https://img.limour.top/inload.svg" lazyload alt="" style="width:auto;height:24px"></a><br><a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/limour-blog/limour-blog.github.io" target="_blank" rel="nofollow noopener"><span>SRC</span></a> <i class="iconfont icon-love"></i> <a href="https://web.archive.org/web/20231130095837/https://effectiveacceleration.tech/" target="_blank" rel="nofollow noopener"><span>e/Acc</span></a></div></div></footer><script src="/cdn/nprogress/nprogress.min.js"></script><link rel="stylesheet" href="/cdn/nprogress/nprogress.min.css"><script>NProgress.configure({showSpinner:!1,trickleSpeed:100}),NProgress.start(),window.addEventListener("load",function(){NProgress.done()})</script><script src="/cdn/jquery/jquery.min.js"></script><script src="/cdn/bootstrap/js/bootstrap.min.js"></script><script src="/js/events.js"></script><script src="/js/plugins.js"></script><script src="/js/img-lazyload.js"></script><script>Fluid.utils.createScript("/cdn/tocbot/tocbot.min.js",function(){var t,o=jQuery("#toc");0!==o.length&&window.tocbot&&(t=jQuery("#board-ctn").offset().top,window.tocbot.init(Object.assign({tocSelector:"#toc-body",contentSelector:".markdown-body",linkClass:"tocbot-link",activeLinkClass:"tocbot-active-link",listClass:"tocbot-list",isCollapsedClass:"tocbot-is-collapsed",collapsibleClass:"tocbot-is-collapsible",scrollSmooth:!0,includeTitleTags:!0,headingsOffset:-t},CONFIG.toc)),0<o.find(".toc-list-item").length&&o.css("visibility","visible"),Fluid.events.registerRefreshCallback(function(){var t;"tocbot"in window&&(tocbot.refresh(),0!==(t=jQuery("#toc")).length)&&tocbot&&0<t.find(".toc-list-item").length&&t.css("visibility","visible")}))})</script><script src="/cdn/clipboard-js/clipboard.min.js"></script><script>Fluid.plugins.codeWidget()</script><script>Fluid.utils.createScript("/cdn/anchor-js/anchor.min.js",function(){window.anchors.options={placement:CONFIG.anchorjs.placement,visible:CONFIG.anchorjs.visible},CONFIG.anchorjs.icon&&(window.anchors.options.icon=CONFIG.anchorjs.icon);var n,o=[];for(n of(CONFIG.anchorjs.element||"h1,h2,h3,h4,h5,h6").split(","))o.push(".markdown-body > "+n.trim());"left"===CONFIG.anchorjs.placement&&(window.anchors.options.class="anchorjs-link-left"),window.anchors.add(o.join(", ")),Fluid.events.registerRefreshCallback(function(){if("anchors"in window){anchors.removeAll();var n,o=[];for(n of(CONFIG.anchorjs.element||"h1,h2,h3,h4,h5,h6").split(","))o.push(".markdown-body > "+n.trim());"left"===CONFIG.anchorjs.placement&&(anchors.options.class="anchorjs-link-left"),anchors.add(o.join(", "))}})})</script><script>Fluid.plugins.imageCaption()</script><script src="/js/local-search.js"></script><script src="/js/boot.js"></script><noscript><div class="noscript-warning">Blog works best with JavaScript enabled</div></noscript><script defer src="/theme-inject/timeliness.js"></script></body></html>